components/brave_wallet/browser/json_rpc_service_unittest.cc

```diff
@@ -7488,4 +7509,129 @@ TEST_F(JsonRpcServiceUnitTest, AnkrGetAccountBalances) {
+  ASSERT_TRUE(AddUserAsset(prefs(), asset.Clone()));
+
+  auto asset2 = mojom::BlockchainToken::New(
+      "So11111111111111111111111111111111111111112", "Wrapped SOL", "sol.png",
+      false, false, false, mojom::SPLTokenProgram::kUnknown, false, false,
+      "WSOL", 8, true, "", "", mojom::kSolanaMainnet, mojom::CoinType::SOL);
+  ASSERT_TRUE(AddUserAsset(prefs(), asset2.Clone()));
+
+  // Test record in registry, the value should be used.
+  TestGetSPLTokenProgramByMint(
+      FROM_HERE, "2kMpEJCZL8vEDZe7YPLMCS9Y3WKSAMedXBn7xHPvsWvi",
+      mojom::kSolanaMainnet, mojom::SPLTokenProgram::kToken2022,
+      mojom::SolanaProviderError::kSuccess, "");
+
+  // Test record in both registry and user assets. The value in user assets
+  // should be used.
+  TestGetSPLTokenProgramByMint(FROM_HERE, tsla_mint_addr, mojom::kSolanaMainnet,
+                               mojom::SPLTokenProgram::kToken2022,
+                               mojom::SolanaProviderError::kSuccess, "");
+
+  std::string json = R"(
+    {
+      "jsonrpc":"2.0","id":1,
+      "result": {
+        "context":{"slot":123065869},
+        "value":{
+          "data":["SEVMTE8gV09STEQ=","base64"],
+          "executable":false,
+          "lamports":18446744073709551615,
+          "owner":"$1",
+          "rentEpoch":18446744073709551615
+        }
+      }
+    }
+  )";
+
+  // Test record in user assets with unknown token program, result is from
+  // network and the pref value should be updated based on the result.
+  auto user_asset =
+      GetUserAsset(prefs(), mojom::CoinType::SOL, mojom::kSolanaMainnet,
+                   asset2->contract_address, "", false, false);
+  ASSERT_TRUE(user_asset);
+  EXPECT_EQ(user_asset->spl_token_program, mojom::SPLTokenProgram::kUnknown);
+
+  auto expected_network_url =
+      GetNetwork(mojom::kSolanaMainnet, mojom::CoinType::SOL);
+  SetInterceptor(expected_network_url, "getAccountInfo", "",
+                 base::ReplaceStringPlaceholders(
+                     json, {mojom::kSolanaTokenProgramId}, nullptr));
+
+  TestGetSPLTokenProgramByMint(
+      FROM_HERE, asset2->contract_address, mojom::kSolanaMainnet,
+      mojom::SPLTokenProgram::kToken, mojom::SolanaProviderError::kSuccess, "");
+
+  user_asset =
+      GetUserAsset(prefs(), mojom::CoinType::SOL, mojom::kSolanaMainnet,
+                   asset2->contract_address, "", false, false);
+  ASSERT_TRUE(user_asset);
+  EXPECT_EQ(user_asset->spl_token_program, mojom::SPLTokenProgram::kToken);
+
+  // Test record not in registry or user assets, result is from network.
+  SetInterceptor(expected_network_url, "getAccountInfo", "",
+                 base::ReplaceStringPlaceholders(
+                     json, {mojom::kSolanaToken2022ProgramId}, nullptr));
+  TestGetSPLTokenProgramByMint(
+      FROM_HERE, "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v",
+      mojom::kSolanaMainnet, mojom::SPLTokenProgram::kToken2022,
+      mojom::SolanaProviderError::kSuccess, "");
+
```

Comment:
Shouldn't we also test that result will be updated in user assets by `GetUserAsset`?

Reply:
I'll add comments to test and impl, we only update the pref if there's an existing entry in pref.

Reply:
Actually there's a comment states this above GetSPLTokenProgramByMint in the header.